File: src/go/svd/validators/general_validator.go
package validators

import "github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/models"

// GeneralInformationValidatorImpl validates GeneralInformation.
type GeneralInformationValidatorImpl struct{}

// NewGeneralInformationValidator creates a new GeneralInformation validator.
func NewGeneralInformationValidator() GeneralInformationValidator {
	return &GeneralInformationValidatorImpl{}
}

// Validate validates GeneralInformation using the same rules as the .NET implementation.
func (v *GeneralInformationValidatorImpl) Validate(general *models.GeneralInformation) []ValidationError {
	var errors []ValidationError
	if general == nil {
		return []ValidationError{{
			Field:   "General",
			Message: "General is required",
		}}
	}

	if general.Imo == "" {
		errors = append(errors, ValidationError{Field: "General.Imo", Message: "Imo is required", Value: general.Imo})
	}
	if !isSevenDigitImo(general.Imo) {
		errors = append(errors, ValidationError{Field: "General.Imo", Message: "Imo must be seven digits.", Value: general.Imo})
	}
	if general.ShipName == "" {
		errors = append(errors, ValidationError{Field: "General.ShipName", Message: "Ship Name is required", Value: general.ShipName})
	}
	if general.ShipReportingDate.IsZero() {
		errors = append(errors, ValidationError{Field: "General.ShipReportingDate", Message: "Ship Reporting Date (Datetime) must be greater than default.", Value: general.ShipReportingDate})
	}

	return errors
}

func isSevenDigitImo(value string) bool {
	if len(value) != 7 {
		return false
	}

	for i := 0; i < len(value); i++ {
		if value[i] < '0' || value[i] > '9' {
			return false
		}
	}

	return true
}

File: src/typescript/src/validators/GeneralValidator.ts
import { asObject, readProperty, readTrimmedString } from "../runtime/types";
import { ValidationError } from "./ValidationError";

export class GeneralValidator {
  public validate(general: unknown): ValidationError[] {
    const generalObject = asObject(general);
    if (generalObject === null) {
      return [{ field: "General", message: "General is required", value: general }];
    }

    const errors: ValidationError[] = [];
    const imo = readTrimmedString(generalObject, "imo");
    const shipName = readTrimmedString(generalObject, "shipName");
    const shipReportingDate = readProperty(generalObject, "shipReportingDate");

    if (imo.length === 0) {
      errors.push({
        field: "General.Imo",
        message: "Imo is required",
        value: imo,
      });
    }

    if (!/^\d{7}$/.test(imo)) {
      errors.push({
        field: "General.Imo",
        message: "Imo must be seven digits.",
        value: imo,
      });
    }

    if (shipName.length === 0) {
      errors.push({
        field: "General.ShipName",
        message: "Ship Name is required",
        value: shipName,
      });
    }

    if (!isValidShipReportingDate(shipReportingDate)) {
      errors.push({
        field: "General.ShipReportingDate",
        message: "Ship Reporting Date (Datetime) must be greater than default.",
        value: shipReportingDate,
      });
    }

    return errors;
  }
}

function isValidShipReportingDate(value: unknown): boolean {
  const parsed = parseDate(value);
  if (parsed === null) {
    return false;
  }

  return parsed.getUTCFullYear() > 1;
}

function parseDate(value: unknown): Date | null {
  if (value instanceof Date) {
    return Number.isNaN(value.valueOf()) ? null : value;
  }

  if (typeof value !== "string") {
    return null;
  }

  const trimmed = value.trim();
  if (trimmed.length === 0) {
    return null;
  }

  const parsed = new Date(trimmed);
  if (Number.isNaN(parsed.valueOf())) {
    return null;
  }

  return parsed;
}
